tools/cli/src/cli/common.rs

```diff
@@ -101,19 +101,41 @@ impl CommonOpt {
             return Err(ToClientError::Other);
         }
 
-        // we need to store guessed username for login and reauth.
-        let username;
-
         // If we have a username, use that to select tokens
-        let token = match &self.username {
-            Some(_username) => {
-                username = _username.clone();
+        let (spn, token) = match &self.username {
+            Some(filter_username) => {
+                let possible_token = if filter_username.contains('@') {
+                    // If there is an @, it's an spn so just get the token directly.
+                    tokens
+                        .get(filter_username)
+                        .map(|t| (filter_username.clone(), t.clone()))
+                } else {
+                    // First, filter for tokens that match.
+                    let mut token_refs: Vec<_> = tokens
+                        .iter()
+                        .filter(|(t, _)| t.starts_with(filter_username))
```

Comment:
Shouldn't this have an @ appended to the end? If you do `-D foo` and the token `foobar@example.com` exists, that'll match.

```suggestion
                        .filter(|(t, _)| t.starts_with(&format!("{}@", filter_username)))
```